### PR TITLE
166791126 update and delete author

### PR DESCRIPTION
**Pull request by JuwonAbiola**
### What does this PR do?
Add update and delete author

#### Description of Task to be completed?
- Create new endpoints to edit and delete author
- Write unit test to test all routers

#### How should this be manually tested?
- Clone or download this branch
- run npm install to  Install all dependencies
- Run npm run db:rollback &&  npm run start:dev
- Open postman use: 
       update or delete
       http://localhost:{PORT}/api/v1/admin/author/:authorId
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#166791126](https://www.pivotaltracker.com/n/projects/2350588/stories/166791126)
#### Screenshots (if appropriate)
image
image
image

#### Questions:
N/A

### CODE REVIEW

**Comment by allebd** [comment]
No screenshots attached

**Comment by allebd** [comment]
No swagger documentation

**Review by allebd — COMMENTED** [review]
I observed there was no documentation and no screenshots in PR

**Comment by Lundii** [comment]
> I observed there was no documentation and no screenshots in PR

Yeah. Please add the swagger docs and a screenshot to your PR

File: server/controllers/AuthorController.js [hunk 1]
```diff
@@ -29,4 +29,46 @@ export default class AuthorController {
...
+   */
+  static async updateAuthor(request, response) {
+    const { params: { authorId }, body: { authorName } } = request;
+    try {
+      const existingAuthor = await Author.findOne({ where: { id: authorId } });
+      if (!existingAuthor) { return responseMessage(response, 404, { message: 'author not found' }); }
+      const sameAuthor = await Author.findOne({ where: { fullname: authorName } });
+      if (sameAuthor) { return responseMessage(response, 409, { message: 'author already exist' }); }
+      await Author.update({ fullname: authorName }, { where: { id: authorId } });
+      return responseMessage(response, 200, { status: 'success', message: 'author updated successfully' });
```
Comment (allebd) [inline server/controllers/AuthorController.js]: No data response of the updated author

Comment (JuwonAbiola) [reply server/controllers/AuthorController.js]: done

File: server/controllers/AuthorController.js [hunk 2]
```diff
@@ -29,4 +29,50 @@ export default class AuthorController {
...
+   * @memberof AuthorControllers
+   */
+  static async updateAuthor(request, response) {
+    const { params: { authorId }, body: { authorName } } = request;
+    try {
+      const existingAuthor = await Author.findOne({ where: { id: authorId } });
+      if (!existingAuthor) { return responseMessage(response, 404, { message: 'author not found' }); }
+      const sameAuthor = await Author.findOne({ where: { fullname: authorName } });
+      if (sameAuthor) { return responseMessage(response, 409, { message: 'author already exist' }); }
+      const authorUpdate = await Author.update({ fullname: authorName }, {
```
Comment (Nennyfills) [inline server/controllers/AuthorController.js:49]: you have to also update the updatedAt too

File: server/controllers/AuthorController.js [hunk 3]
```diff
@@ -29,4 +29,50 @@ export default class AuthorController {
...
+   * @param {object} response express response object
+   * @returns {json} json
+   * @memberof AuthorControllers
+   */
+  static async deleteAuthor(request, response) {
+    const { params: { authorId } } = request;
+    try {
+      const existingAuthor = await Author.findOne({ where: { id: authorId } });
+      if (!existingAuthor) { return responseMessage(response, 404, { message: 'author not found' }); }
+      Author.destroy({ where: { id: authorId } });
```
Comment (Nennyfills) [inline server/controllers/AuthorController.js]: is this not supposed to have await in front of it?

File: test/server/routes/admin.test.js [hunk 4]
```diff
@@ -328,7 +332,243 @@ describe('ADMIN ROUTES', () => {
...
+          expect(response.body.message).to.be.equal('user successfully activated');
+          done();
+        });
+    });
+    it('returns a status 400 if user is already inactive', (done) => {
+      const body = {
+        status: 'inactive'
+      };
+      chai.request(app)
+        .patch('/api/v1/admin/user/8')
```
Comment (Nennyfills) [inline test/server/routes/admin.test.js:406]: why not declare your route once, to avoid duplication

File: server/controllers/AuthorController.js [hunk 5]
```diff
@@ -29,4 +29,50 @@ export default class AuthorController {
...
+   * @param {object} response express response object
+   * @returns {json} json
+   * @memberof AuthorControllers
+   */
+  static async updateAuthor(request, response) {
+    const { params: { authorId }, body: { authorName } } = request;
+    try {
+      const existingAuthor = await Author.findOne({ where: { id: authorId } });
+      if (!existingAuthor) { return responseMessage(response, 404, { message: 'author not found' }); }
+      const sameAuthor = await Author.findOne({ where: { fullname: authorName } });
```
Comment (Nennyfills) [inline server/controllers/AuthorController.js]: This const existingAuthor = await Author.findOne({ where: { id: authorId } }); and
const sameAuthor = await Author.findOne({ where: { fullname: authorName } }); are thesame,
why not get all author with  Author.findAll(), and then use map and find to get the author you want, insteading constantly querying the database.

**Review by Nennyfills — DISMISSED** [review]

File: swagger.json [hunk 6]
```diff
@@ -615,57 +743,7 @@
         } ]
       }
     }
-  },
-"/books": {
```
Comment (Lundii) [inline swagger.json]: Why did you delete all these?

Comment (Lundii) [inline swagger.json]: Why did you delete this route?

**Review by Lundii — DISMISSED** [review]

**Review by Adekoreday — DISMISSED** [review]

**Review by henryade — DISMISSED** [review]

File: server/controllers/AuthorController.js (hunk 5 above)
Comment (Ukhu) [inline server/controllers/AuthorController.js]: The logic here is quite wrong. Why query the database to check if that name already exists? The author's fullname isn't a unique attribute, hence, it is possible to actually have two authors with the same name but different IDs. What you should rather do is to check if the fullname of the author with the given authorId is the same as the authorName in the params. This way, the response will tell the user that the fullname for the author with that ID is already the same as the authorName being passed in.

**Review by Ukhu — CHANGES_REQUESTED** [review]

File: server/controllers/AuthorController.js (hunk 5 above)
Comment (JuwonAbiola) [reply server/controllers/AuthorController.js]: done

Comment (allebd) [reply server/controllers/AuthorController.js]: @Ukhu, author's fullname is meant to be unique, why would an author have two ids

Comment (Ukhu) [reply server/controllers/AuthorController.js]: I didn't say one author should have two IDs. What I am trying to say is that 2 different authors can have the same. There can be two authors named Bella Oyedele. How do you differentiate them? That's why the ID is important. Author fullname can't be unique.

File: server/controllers/AuthorController.js [hunk 7]
```diff
@@ -29,4 +29,52 @@ export default class AuthorController {
...
+   * @description allows admin to update an author
+   * @param {object} request express request object
+   * @param {object} response express response object
+   * @returns {json} json
+   * @memberof AuthorControllers
+   */
+  static async updateAuthor(request, response) {
+    const { params: { authorId }, body: { authorName } } = request;
+    try {
+      const authors = await Author.findAll();
```
Comment (Ukhu) [inline server/controllers/AuthorController.js:43]: I prefer the way you did it before. There is no need to get all the authors from the database at once. If there are very many authors in the database, this query would be very intensive. Use the findOne instead. That is better.

**Review by Ukhu — DISMISSED** [review]

**Review by henryade — DISMISSED** [review]

**Review by Adekoreday — DISMISSED** [review]

**Review by allebd — DISMISSED** [review]

**Review by henryade — APPROVED** [review]

**Review by Adekoreday — APPROVED** [review]

**Review by allebd — APPROVED** [review]

**Review by Ukhu — APPROVED** [review]
